### PR TITLE
task framework idea

### PR DESCRIPTION
**Pull request by Bryce-Soghigian**

### CODE REVIEW

File: bryce-task-framework.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+we know the topological order of execution needs to be
+
+```
+super_dependency: 0
+subtask_two: 1
+subtask_one: 1
+task: 2
+```
+We infer this from the yaml. Now we need to execute the tasks in this order, but you notice we have two tasks which are the same. For these tasks we can execute them concurrently.
+So we first load the executables of the lowest indegree, then we execute them. But we have a problem. At the higher level we have a schedule of every 300 seconds. what if the execution of these
```
Comment (Takashiidobe) [inline bryce-task-framework.md:105]: Ah, well done, I didn't think of this. In my solution I didn't address this.

File: bryce-task-framework.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+def get_status(status: int):
+	if status == 0:
+		return "CREATED"
+	elif status == 1:
+		return "ENQUEUED"
+	elif status == 2:
+		return "STARTED"
+	else:
+		return "FAILURE"
```
Comment (Takashiidobe) [inline bryce-task-framework.md:134]: I'd prefer this:
The reason why is that this can provide incorrect information for a mix of new and old clients.

Imagine you have some nodes that are running a previous version of your code. Now, you want to add a new status, like "DEPENDENCY_FAILURE" if in case you want to bubble up to a user that the topological sort failed. 

You'll set this to the value 4 (since I assume FAILURE is 3).

If there are nodes running the previous version of the code, they'll interpret it as "3", or FAILURE, which is incorrect. If we raise an unimplemented error in case we pass the old code a "4", we won't give the client incorrect information.

```suggestion
def get_status(status: int):
	if status == 0:
		return "CREATED"
	elif status == 1:
		return "ENQUEUED"
	elif status == 2:
		return "STARTED"
	elif status == 3:
		return "FAILURE"
	else:
	        raise UnImplementedError("Unimplemented")
```

File: bryce-task-framework.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+```
+
+A given task will be enqueued in execution in celery pulling nodes from redis, then we update some elasticsearch instance for each node.
+
+```python
+@dataclass
+class TaskNode:
+	status: int
+	task_id: uuid
```
Comment (Takashiidobe) [inline bryce-task-framework.md:143]: Given that we have a billion rows+ every day, i think a uuid is going to cause you a bit of trouble.
UUIDs aren't sorted, so even though you can index them with a b-tree or a hash-index, they can be slow:

https://www.percona.com/blog/2019/11/22/uuids-are-popular-but-bad-for-performance-lets-discuss/

You can, however, increase throughput considerably using a sequentialish uuid w/ the timestamp encoded:

https://www.2ndquadrant.com/en/blog/sequential-uuid-generators/

Basically, databases really love sequential data, and performance can vary quite considerably if you use a uuid vs a plain int. 

Here I would ask myself if it's worth it to bottleneck at the database level given the amounts of writes i'm doing, by allow sequential id generation vs taking the performance hit (inserts can be slower with a uuid, and indexes are much bigger: https://www.cybertec-postgresql.com/en/int4-vs-int8-vs-uuid-vs-numeric-performance-on-bigger-joins/)

File: bryce-task-framework.md (hunk 2 above)
Comment (Bryce-Soghigian) [reply bryce-task-framework.md:134]: python has a NotImplementedError Exception that could be used here

File: bryce-task-framework.md (hunk 1 above)
Comment (Bryce-Soghigian) [reply bryce-task-framework.md:105]: im good at the algorithms based stuff but need to improve everywhere else in my design.

**Review by Takashiidobe — APPROVED** [review]
